flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/DenseMatrix.java

```diff
@@ -0,0 +1,751 @@
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.matrix;
+
+import java.io.Serializable;
+import java.util.Arrays;
+
+/**
+ * Dense Matrix.
+ */
+public class DenseMatrix implements Serializable {
+	/* ------------------------
+	Class variables
+     * ------------------------ */
+
+	/**
+	 * Row and column dimensions.
+	 */
+	int m, n;
+
+	/**
+	 * Array for internal storage of elements.
+	 * <p>
+	 * The matrix data is stored in column major format internally.
+	 */
+	double[] data;
+
+    /* ---------------------------------------------------
+	 * Constructors
+     * --------------------------------------------------- */
+
+	/**
+	 * Construct an empty matrix.
+	 */
+	public DenseMatrix() {
+	}
+
+	/**
+	 * Construct an m-by-n matrix of zeros.
+	 *
+	 * @param m Number of rows.
+	 * @param n Number of colums.
+	 */
+	public DenseMatrix(int m, int n) {
+		this.m = m;
+		this.n = n;
+		this.data = new double[m * n];
+	}
+
+	/**
+	 * Construct a matrix from a 1-D array. The data in the array should organize
+	 * in row major.
+	 *
+	 * @param m    Number of rows.
+	 * @param n    Number of cols.
+	 * @param data One-dimensional array of doubles.
+	 */
+	public DenseMatrix(int m, int n, double[] data) {
+		this(m, n, data, true);
+	}
+
+	/**
+	 * Construct a matrix from a 1-D array. The data in the array is organized
+	 * in column major or in row major, which is specified by parameter 'inRowMajor'
+	 *
+	 * @param m          Number of rows.
+	 * @param n          Number of cols.
+	 * @param data       One-dimensional array of doubles.
+	 * @param inRowMajor Whether the matrix in 'data' is in row major format.
+	 */
+	public DenseMatrix(int m, int n, double[] data, boolean inRowMajor) {
+		assert (data.length == m * n);
+		this.m = m;
+		this.n = n;
+		if (inRowMajor) {
+			this.data = new double[m * n];
+			for (int i = 0; i < m; i++) {
+				for (int j = 0; j < n; j++) {
+					this.set(i, j, data[i * n + j]);
+				}
+			}
+		} else {
+			this.data = data.clone();
+		}
+	}
+
+	/**
+	 * Construct a matrix from a 2-D array.
+	 *
+	 * @param matA Two-dimensional array of doubles.
+	 * @throws IllegalArgumentException All rows must have the same size
+	 */
+	public DenseMatrix(double[][] matA) {
+		this.m = matA.length;
+		this.n = matA[0].length;
+		for (int i = 0; i < m; i++) {
+			if (matA[i].length != n) {
+				throw new IllegalArgumentException("All rows must have the same size.");
+			}
+		}
+		this.data = new double[m * n];
+		for (int i = 0; i < m; i++) {
+			for (int j = 0; j < n; j++) {
+				this.set(i, j, matA[i][j]);
+			}
+		}
+	}
+
+	/**
+	 * Construct a matrix with provided data buffer.
+	 * This is for internal use only, so it is package private.
+	 *
+	 * @param m    Number of rows.
+	 * @param n    Number of cols.
+	 * @param data One-dimensional array of doubles.
+	 */
+	public static DenseMatrix fromDataBuffer(int m, int n, double[] data) {
+		assert m * n == data.length;
+		DenseMatrix matA = new DenseMatrix();
```

Comment:
There is already a similar constructor where the double array is cloned. However, this interface is intended for performance critical cases by not cloning the array. We have moved this interface to DenseMatrixUtil to avoid ambiguities.